cedar-policy-validator/src/types.rs

```diff
@@ -614,6 +625,109 @@ impl Type {
+                    },
+                    None => Ok(false),
+                }
+            }
+            Type::EntityOrRecord(EntityRecordKind::ActionEntity { name, .. }) => {
+                match restricted_expr.as_euid() {
+                    Some(euid) if euid.is_action() => match euid.entity_type() {
+                        EntityType::Concrete(euid_name) => Ok(euid_name == name),
+                        EntityType::Unspecified => Ok(false), // `euid.is_action()` should never be true for an Unspecified, so this should be unreachable, but it's safe to return Ok(false), because Unspecified can only have the validator type `AnyEntity` anyway
+                    },
+                    _ => Ok(false),
+                }
+            }
+            Type::EntityOrRecord(EntityRecordKind::AnyEntity) => {
+                Ok(restricted_expr.as_euid().is_some())
+            }
+            Type::EntityOrRecord(EntityRecordKind::Record {
+                attrs,
+                open_attributes,
+            }) => match restricted_expr.as_record_pairs() {
+                Some(pairs) => {
+                    let record: HashMap<_, BorrowedRestrictedExpr<'_>> = pairs.collect();
+                    for (k, attr_val) in &record {
+                        match attrs.get_attr(&k) {
+                            Some(attr_ty) => {
+                                if !attr_ty
+                                    .attr_type
+                                    .typecheck_restricted_expr(attr_val.to_owned(), extensions)?
+                                {
+                                    return Ok(false);
+                                }
+                            }
+                            None => {
+                                if open_attributes != &OpenTag::OpenAttributes {
+                                    // the restricted expr has an attribute not
+                                    // listed in the Type, and the Type doesn't
+                                    // have open attributes
+                                    return Ok(false);
+                                }
+                            }
+                        }
+                    }
+                    // we've now checked that all of the attrs in `restricted_expr` are OK and have the right types.
+                    // what remains is making sure that all the required attrs are actually in `restricted_expr`
+                    for (k, attr_ty) in attrs.iter() {
+                        if attr_ty.is_required && !record.contains_key(k) {
+                            return Ok(false);
+                        }
+                    }
+                    Ok(true)
+                }
+                None => Ok(false),
+            },
+            Type::ExtensionType { name } => match restricted_expr.as_extn_fn_call() {
```

Comment:
Does this not need to check the arguments? I suppose that doesn't effect the type of the expression, assuming it is well typed, but it's probably good to reject a request with a bad extension function call.

Reply:
Fixed in 8908ee5